pkg/util/voljournal.go

```diff
@@ -0,0 +1,383 @@
+}
+
+/*
+UndoReservation undoes a reservation, in the reverse order of ReserveName
+- The UUID directory is cleaned up before the VolName key in the csiDirectory is cleaned up
+
+NOTE: Ensure that the Ceph volume (image or FS subvolume) backing the reservation is cleaned up
+prior to cleaning up the reservation
+
+NOTE: As the function manipulates omaps, it should be called with a lock against the request name
+held, to prevent parallel operations from modifying the state of the omaps for this request name.
+*/
+func (cj *CSIJournal) UndoReservation(monitors, id, key, pool, volName, reqName string) (err error) {
+	// delete volume UUID omap (first, inverse of create order)
+	// TODO: Check cases where volName can be empty, and we need to just cleanup the reqName
+	imageUUID := strings.TrimPrefix(volName, cj.namingPrefix)
+	err = RemoveObject(monitors, id, key, pool, cj.cephUUIDDirectoryPrefix+imageUUID)
+	if err != nil {
+		if _, ok := err.(ErrObjectNotFound); !ok {
+			klog.Errorf("failed removing oMap %s (%s)", cj.cephUUIDDirectoryPrefix+imageUUID, err)
+			return err
+		}
+	}
+
+	// delete the request name key (last, inverse of create order)
+	err = RemoveOMapKey(monitors, id, key, pool, cj.csiDirectory,
+		cj.csiNameKeyPrefix+reqName)
+	if err != nil {
+		klog.Errorf("failed removing oMap key %s (%s)", cj.csiNameKeyPrefix+reqName, err)
+		return err
+	}
+
+	return nil
+}
+
+// reserveOMapName creates an omap with passed in oMapNamePrefix and a generated <uuid>.
+// It ensures generated omap name does not already exist and if conflicts are detected, a set
+// number of retires with newer uuids are attempted before returning an error
+func reserveOMapName(monitors, id, key, pool, oMapNamePrefix string) (string, error) {
+	var iterUUID string
+
+	maxAttempts := 5
+	attempt := 1
+	for attempt <= maxAttempts {
+		// generate a uuid for the image name
+		iterUUID = uuid.NewUUID().String()
+
+		err := CreateObject(monitors, id, key, pool, oMapNamePrefix+iterUUID)
+		if err != nil {
+			if _, ok := err.(ErrObjectExists); ok {
+				attempt++
+				// try again with a different uuid, for maxAttempts tries
+				klog.V(4).Infof("uuid (%s) conflict detected, retrying (attempt %d of %d)",
+					iterUUID, attempt, maxAttempts)
+				continue
+			}
+
+			return "", err
+		}
+
+		break
+	}
+
+	if attempt > maxAttempts {
+		return "", errors.New("uuid conflicts exceeds retry threshold")
+	}
+
+	return iterUUID, nil
+}
+
+/*
+ReserveName adds respective entries to the csiDirectory omaps, post generating a target
+UUIDDirectory for use. Further, these functions update the UUIDDirectory omaps, to store back
+pointers to the CSI generated request names.
+
+NOTE: As the function manipulates omaps, it should be called with a lock against the request name
+held, to prevent parallel operations from modifying the state of the omaps for this request name.
+*/
+func (cj *CSIJournal) ReserveName(monitors, id, key, pool, reqName, parentName string) (string, error) {
```

Comment:
Just a thought:-

here named return is not used, can we stick to one approach?
can we follow this approach in other places too?

Reply:
Done (removed any further instances of naming return vars as well)